volatility/framework/symbols/windows/extensions/network.py

```diff
@@ -0,0 +1,297 @@
+        # If we start or end with None, then add an additional :
+        if words[0] is None:
+            words = [None] + words
+        if words[-1] is None:
+            words += [None]
+        # Join up everything we've got using :s
+        return ":".join(["{0:x}".format(w) if w is not None else "" for w in words]) + encapsulated
+
+    if address_family == socket.AF_INET:
+        return inet_ntop4(packed_ip)
+    elif address_family == socket.AF_INET6:
+        return inet_ntop6(packed_ip)
+    raise socket.error("[Errno 97] Address family not supported by protocol")
+
+# Python's socket.AF_INET6 is 0x1e but Microsoft defines it
+# as a constant value of 0x17 in their source code. Thus we
+# need Microsoft's since that's what is found in memory.
+AF_INET = 2
+AF_INET6 = 0x17
+
+# String representations of INADDR_ANY and INADDR6_ANY
+inaddr_any = inet_ntop(socket.AF_INET, [0] * 4)
+inaddr6_any = inet_ntop(socket.AF_INET6, [0] * 16)
+
+# copied to here as symbol space does not appear to support custom Enums
+TCP_STATE_ENUM = {
+    0: 'CLOSED',
+    1: 'LISTENING',
+    2: 'SYN_SENT',
+    3: 'SYN_RCVD',
+    4: 'ESTABLISHED',
+    5: 'FIN_WAIT1',
+    6: 'FIN_WAIT2',
+    7: 'CLOSE_WAIT',
+    8: 'CLOSING',
+    9: 'LAST_ACK',
+    12: 'TIME_WAIT',
+    13: 'DELETE_TCB'
+}
+
+class _TCP_LISTENER(objects.StructType):
+    """Class for objects found in TcpL pools.
+
+    This class serves as a base class for all pooled network objects.
+
+    It exposes some functions which return sanity-checked members. Substructures referred to by a
+    pointer may appear valid at first glance but will throw an InvalidAddressException on access.
+
+    This is not a problem when objects are validated via their `is_valid()` method, but when
+    scanning for semi-corrupted data this check will not be performed.
+
+    Be mindful that most of those methods return `None` when they would access invalid data.
+    If you want to process the raw data access the attributes directly, e.g.
+    via `network_object.InetAF` instead of `network_object.get_address_family()`.
+
+    """
+
+    def __init__(self, context: interfaces.context.ContextInterface, type_name: str,
+                 object_info: interfaces.objects.ObjectInformation, size: int,
+                 members: Dict[str, Tuple[int, interfaces.objects.Template]]) -> None:
+
+        super().__init__(context = context,
+                         type_name = type_name,
+                         object_info = object_info,
+                         size = size,
+                         members = members)
+
+    def get_address_family(self):
+        try:
+            return self.InetAF.dereference().AddressFamily
+
+        except exceptions.InvalidAddressException:
+            return None
+
+    def get_owner(self):
+        try:
+            return self.member('Owner').dereference()
+
+        except exceptions.InvalidAddressException:
+            return None
+
+    def get_owner_pid(self):
```

Comment:
Hmmm, undecided on how I feel about bubbling up child attributes in order to put the error checking in a single place.  5:S  It's probably the best way of doing it, but I'm almost wondering if it's worth an adaptor you can sit on top of an object that checks validity and returns None for an attribute if it would otherwise fail?

Something like:
`safe_owner = SafePointerAccess(self.get_owner())`
and then `safe_owner` could do the is_valid check on construction, and also check `has_valid_member`, so that all the other calls could become `safe_owner.pid`, etc?  Any thoughts?  @japhlange @atcuno

Reply:
That would require all attributes wrapped within a `SafePointerAccess` to possess an `is_valid()` method, right? (Or at least to wrap their validation within a function of that name, if existent.)
In the end I think this is a design question. While I agree that a `try/catch` per attribute per object clutters the code and makes it longer it _could_ make it a bit easier to understand on first read. Otherwise you'd need to look into another abstraction layer.
On the other hand this abstraction might be worthwhile if you would need to use several ways to validate the object, e.g. `is_valid()`, `is_valid_member()` and maybe some additional semantic checks as well. If those got wrapped within one adaptor it could actually help readability. To abstract or not to abstract? :)
In this case we would change from
```python
def get_owner_pid(self):
    try:
        if self.get_owner().is_valid():
            return self.get_owner().UniqueProcessId
        else:
            return None
    except exceptions.InvalidAddressException:
        return None
```
via
```python
def get_owner_pid(self):
    if self.get_owner().is_valid():
        if self.get_owner().has_valid_member("UniqueProcessId"):
            return self.get_owner().UniqueProcessId
    return None
```
to this?
```python
def get_owner_pid(self):
    safe_owner = SafePointerAccess(self.get_owner())
    return safe_owner.UniqueProcessId # returns either int or None
```

Reply:
But then this still begs the question what to do with the returned `None` - at some point in the interface it has to be swapped out with a `renderers.UnreadableValue()`.

Reply:
So I was thinking more along the lines of:
```
    safe_owner = SafePointerAccess(self, on_fail_returns = renderers.UnreadableValue()).get_owner()
```
But you're right, it does need to end up as the right value by the end.  I'm still not convinced None is the right value at the moment.  Recall, it wouldn't need to be in its own separate method, it could just be used where it's needed, so it should reduce a lot of proxied child calls, but I admit, it may hard readability...

Reply:
If the adaptor returns `UnreadableValue()` this will also be returned when called by a different plugin, so that plugin would need to catch that value.

Reply:
Still thinking about this, but it's not a blocker.